torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -28,9 +36,141 @@ def __init__(
+            return  # no-op
+        self._all_gather_result = foreach_all_gather(
+            self.fsdp_params,
+            self._all_gather_process_group,
+            async_op,
+            self._all_gather_copy_in_stream_for_unshard,
+            self._all_gather_stream_for_unshard,
+            self.device,
+            self._param_dtype,
+        )
+
+    def wait_for_unshard(self):
+        """
+        1. In forward with implict prefetching, to overlap the current copy-out
+        with the next all-gather, we save a reference to the current all-gather
+        result to free after the next copy-out.
+        2. Otherwise (explicit prefetching or in backward), we free the
+        all-gather result immediately after the current copy-out since we can
+        already overlap the current copy-out with the previous reduce-scatter.
+        """
+        if not self._all_gather_result:
+            return  # no preceding unshard
+        if self._training_state == TrainingState.FORWARD:  # implicit prefetch
+            if prev_all_gather_state := self.all_gather_state.pop():
+                self._wait_all_gather_streams_on_event(prev_all_gather_state.event)
+                del prev_all_gather_state  # free
+        foreach_all_gather_copy_out(
+            self._all_gather_result, self.fsdp_params, self._all_gather_process_group
+        )
+        for fsdp_param in self.fsdp_params:
+            fsdp_param.init_unsharded_param()  # no-op after 1st call
+        self._to_unsharded()
+        all_gather_copy_out_event = torch.cuda.Event()
+        all_gather_copy_out_event.record()
+        if self._training_state == TrainingState.FORWARD:
+            self.all_gather_state.put(
+                AllGatherState(self._all_gather_result, all_gather_copy_out_event)
+            )
+        else:
+            self._wait_all_gather_streams_on_event(all_gather_copy_out_event)
+        self._all_gather_result = None  # free unless saved in `all_gather_state`
+
+    def _wait_all_gather_streams_on_event(self, event: torch.cuda.Event):
+        self.all_gather_copy_in_stream.wait_event(event)
+        self.all_gather_stream.wait_event(event)
+
+    def reshard(self):
+        self._to_sharded()
+
+    def pre_forward(
+        self, module: nn.Module, args: Tuple[Any, ...], kwargs: Dict[str, Any]
+    ) -> Tuple[Tuple[Any, ...], Dict[str, Any]]:
+        with torch.profiler.record_function("FSDP::pre_forward"):
+            self._training_state = TrainingState.FORWARD
+            self.unshard()
+            self.wait_for_unshard()
+            return args, kwargs
+
+    def post_forward(self, module: nn.Module, input: Any, output: Any):
+        with torch.profiler.record_function("FSDP::post_forward"):
+            self.reshard()
+            self._training_state = TrainingState.IDLE
+            return output
+
+    # Utilities #
+    def _to_sharded(self):
+        if self._sharded_state != ShardedState.SHARDED:
```

Comment:
Curious if it's valid to call `_to_sharded` in the `ShardedState.SHARDED` state. If not, shall we error out here?

Reply:
It is valid and should be a no-op. For example, for state dict, we want the pre-save and pre-load hooks to call `_to_sharded` to make sure we are exposing the sharded `DTensor` parameters. We prefer to just have the hooks call `_to_sharded` unconditionally.